### ISSUE TITLE
Extension for Procedural textures

### ISSUE THREAD

**Issue by rsahlin** [KhronosGroup/glTF#1889]
Investigating interest in an extension for procedural textures - IKEA_procedural_images  

The purpose of this extension is to enable procedurally generated images that can be used as texture references.  
The proposed solution will offer the possibility of generating textures at load time by defining new mime type(s) and a way of describing image generation from a set of image parameters.  
One goal of this is to reduce transmission size, another goal is to be able to offer dynamic output depending - for instance altering the exact look of a wood texture at loadtime.  

The initial goal is to provide support for wood textures - however the goal is to provide a generic enough descriptive syntax that other types of textures can be created.  

There are two main areas where the glTF spec needs to be updated to support procedural textures:  

## 1: Image reference (mime type)  
For a procedurally generated image to be able to be used as a texture source it must be possible to reference it.  
This could be done by adding one (or more) mime-type - for instance:  
**image/procedural**  

## Procedural generator  
This part will specify some sort of declarative syntax for understanding the data provided to the generator. Input will be one or more datasets - output is the generated image.
It shall be (resonably) easy to implement, provide sufficient performance for real world load time usage and be able to guarantee that the visual output is the same.  
It shall be possible to output different target maps such as color, normal, roughness etc.

**A note on mime-types:**  
As different type of images are needed for different texture channels (color,normal,roughness et) this distinction needs to be done either on mime-type or in texture:
image/procedural/color (for the color image)  
image/procedural/normal (for the normal map)  
image/procedural/orm (for the occlusion, roughness and metallic maps)  

Or, if only one image/procedural mime-type is used the target map needs to be specified in the texture.
While this solution is flexible it will add knowledge about the texture, ie what the target channel is, that is not specified today.
Textures are simply lookups for images using a sampler.
It would perhaps make more sense to add this data to the material - however then the runtime process of generating the target maps could become somewhat convoluted.

**Comment by lexaknyazev** [comment]
While supporting the feature in general, I have to comment on some details.

> This could be done by adding one (or more) mime-type

MIME type is a formally defined term (administered by IANA) designated to identifying specific data formats. glTF's `image` objects are supposed to represent pre-existing resources (hence built-in `bufferView` and `url` options).

To accommodate the possibility of other texture sources, we deliberately made `texture.source` property optional so that extensions on a texture object could provide alternative ways of supplying texture data. FWIW, that property was supposed to be `texture.image` but we didn't manage to rename it in time for glTF 2.0 release.

Assuming that procedural definitions may be reused for different textures, the former should be stored in a new top-level array.

**Comment by rsahlin** [comment]
> Assuming that procedural definitions may be reused for different textures, the former should be stored in a new top-level array.
Thanks for your insights @lexaknyazev - this is way above my level of expertise!

Could you please explain what 'a new top-level array' means and what this would look like?

**Comment by lexaknyazev** [comment]
"A new top-level array" is an array defined within an extension object that is located in the root-level `extensions` slot. For a ratified example, see `KHR_lights_punctual`.

Here's how it may look for procedural textures.
```json
{
  "extensionsUsed": [
    "IKEA_textures_procedural"
  ],
  "asset": {
    "version": "2.0"
  },
  "extensions": {
    "IKEA_textures_procedural": {
      "procedures": [
        {
          "type": "WOOD",
          ...
        }
      ]
    }
  },
  "textures": [
    {
      "extensions": {
        "IKEA_textures_procedural": {
          "procedure": 0
        }
      }
    }
  ]
}
```

**Comment by donmccurdy** [comment]
> This part will specify some sort of declarative syntax for understanding the data provided to the generator...

If we go down the path of supporting procedural textures, we should seriously consider existing standards like MaterialX Standard Nodes:

[MaterialX Specification](https://www.materialx.org/Specification.html)

I'm not suggesting that we include all of MaterialX, here — just to use similar standard nodes for procedural textures, effectively supporting the "pattern graph" subset of MaterialX. This provides very flexible inputs to the existing glTF material system (and upcoming extensions) without defining entirely new shading models. MaterialX can be used with Autodesk Standard Surface — a very good match for glTF's direction — but MaterialX also has additional features for defining custom surface shaders, and I think that should be out of our scope for the time being, as I'm not confident it fits into the realtime transmission ecosystem well right now.

Less impactfully, I'd also suggest `<PREFIX>_texture_<???>` as the name of the extension, to align with https://github.com/KhronosGroup/glTF/tree/master/extensions#naming.

**Comment by rsahlin** [comment]
Thanks @donmccurdy 
I agree - we should try to reuse or be influenced by what is already available in the community.

I guess the first step is to understand what is already available - apart from MaterialX - what are the other (open) ways of procedurally generating textures?

**Comment by rsahlin** [comment]
> "A new top-level array" is an array defined within an extension object that is located in the root-level `extensions` slot. For a ratified example, see `KHR_lights_punctual`.

Ah ok, I think I misunderstood - I thought you where referring to the mime-types.
Yes, sure I see your point.
I'm wondering if the extension should be on the Texture or Image object.

**Comment by donmccurdy** [comment]
NVIDIA's MDL is another good example. My understanding is that it interops well with MaterialX, and maybe even uses the same standard nodes (citation needed?). I'm not aware of any other open, _declarative_ representation of a procedural texture. I say "declarative" because you could argue that OSL, GLSL, etc. are also representations of a procedural texture, but they're not compelling to me for this use case, because they're imperative, less composable, and less portable.

**Comment by donmccurdy** [comment]
>  I'm wondering if the extension should be on the Texture or Image object.

I'd recommend putting the extension on the texture object — this fits very well with how we extend for KTX and WebP textures today.

**Comment by lexaknyazev** [comment]
> I'm wondering if the extension should be on the Texture or Image object.

I'd strongly prefer to keep the semantics of `image` objects intact (i.e. they are simple resource pointers) to not interfere with packaging (think `url`/`bufferView`/base64 conversions) and compression (BasisU, WebP) tools. Keep in mind, that `image.uri` or `image.bufferView` must be always provided per the spec.

**Comment by lexaknyazev** [comment]
> use similar standard nodes for procedural textures, effectively supporting the "pattern graph" subset of MaterialX

The very first step could be even simpler: define and name the most-required texture patterns using something like MaterialX or MDL and just refer to those names from the extension.

**Comment by donmccurdy** [comment]
I suspect that defining high-level texture patterns (wood, checkerboard, etc.) would turn out to be more complicated than defining a discrete set of low-level, composable nodes. MaterialX is able to provide relatively simple OSL reference implementations for each of its standard nodes, which is a big plus to me.

Practically, an engine that supports these high-level texture patterns will absolutely need to assemble shaders at loading time, whether through a node-based system or direct shader generation. If that's the case, I think we should go directly to lower-level nodes as MaterialX, MDL, and USD are doing.

**Comment by meshula** [comment]
FWIW, MatX is higher level than MDL; MatX is a general purpose shader graph, and MDL is specifically a physically based model. MatX will ultimately target MDL, similarly to how it targets OSL and GLSL.

**Comment by pjoe** [comment]
Interesting topic, been thinking about something like this for a while 😄 

One thing that could be worth considering is how these procedurals are created by artist - could this e.g. fit into how Blender3D would do node based procedural textures (I think there is some interest already in blender community about this - also check out this guys stuff: https://gumroad.com/simonthommes).

I suppose Substance Designer is also very much in this space.

**Comment by rsahlin** [comment]
@donmccurdy For our first step in a prototype we would choose one of the 3D tools that support the creation of wood textures and have the code readily available to reverse that 'magic data' into pixels.
We would then use this to turn the procedural data into wood textures at load time.
At this moment in time we are not aiming to solve the nodebased material system.

**Comment by rsahlin** [comment]
> One thing that could be worth considering is how these procedurals are created by artist - could this e.g. fit into how Blender3D would do node based procedural textures (I think there is some interest already in blender community about this - also check out this guys stuff: https://gumroad.com/simonthommes).

This is exactly what we are aiming for in our prototype @pjoe 
We want to condense this down to the absolute minimum (kiss) and start with how the wood texture would be generated in a tool such as Blender.
The next step is then to get hold of the code, or pseudocode, that turns the 'woodnode' into actual pixel data.
As a first step we are not looking into creating a nodebased material system or even generate the wood textures on the fly in the shaders. (though implementations would of course be free to do that if they want to)

**Comment by pjoe** [comment]
Great :)

I do think, like @donmccurdy that you would quickly get to the point where it makes most sense to build this from simpler 'nodes', e.g. 4D noise (which is needed for making procedural noise tiling). IIRC 4D noise nodes landed in blender 2.81.

At load time you could 'execute' the node graph, to generate 'normal' 2D textures (maybe at different resolution depending on device) and hand those of to the actual renderer.

If you haven't already looked at Substance Designer, it is a good source of inspiration for this kind of system - even though it is proprietary.

**Comment by rsahlin** [comment]
> At load time you could 'execute' the node graph, to generate 'normal' 2D textures (maybe at different resolution depending on device) and hand those of to the actual renderer.

Thanks @pjoe - this sounds like what I am after :-)
Do you have any more information regarding how this would be executed from outside Blender for instance?

In order not to choke myself with too much to do I rely on others expertise when it comes to choosing the editing tool :-)
From our perspective the most important thing is to pick one tool for the prototype (eg Blender) and make a proof of concept.
Picking one that is open and has sourcecode readily available is a must.

Later on we will definately look into more complex solutions such as how to define the procedural texture generation.

**Comment by pjoe** [comment]
I don't think Blender currently has the ability to export these procedural texture nodes in any format (except .blend).

You can try asking the Blender community, might be others looking at this (e.g. https://devtalk.blender.org/, https://blender.chat/)

**Comment by donmccurdy** [comment]
> I don't think Blender currently has the ability to export these procedural texture nodes in any format (except .blend).

I've been keeping an eye on things like https://blenderartists.org/t/materialx-blender-integration/700331 and https://github.com/godotengine/godot-proposals/issues/714, but at the moment I don't see any signal from Blender or Godot on how they plan to approach this problem. Blender is working on USD integration, but I'm not sure what that means in terms of node-based materials, if anythingc.

**Comment by rsahlin** [comment]
> For a tightly-controlled pipeline (i.e. you can make the needed changes in 1-2 DCC tools and 1-2 viewers) this should be a reasonable approach. For more widespread adoption I think it would be necessary to shift to lower-level nodes like MaterialX or MDL — I don't expect it is possible to get widespread agreement across the ecosystem on high-level abstractions like "wood".

I agree, our first goal is to do a proof of concept that we are happy with :-)
Step two would be to expand on that and make a solution that declared the wood data in a way the community could embrace.
I fully understand that our fist goal is purely internal - however, I want to get feedback and generate interest for our future approach :-)

**Comment by gustavolsson-ikea** [comment]
> For more widespread adoption I think it would be necessary to shift to lower-level nodes like MaterialX or MDL — I don't expect it is possible to get widespread agreement across the ecosystem on high-level abstractions like "wood".

This is exactly what me and @rsahlin have been discussing. I'm no expert on procedural generation, but I suspect that most procedural representations boil down to a limited number of math operations and noise generation primitives, i.e. a simple math expression and not a turing-complete domain specific language (I do think branching will be required though).

Here are the sources for Blender's texture nodes for example: [https://developer.blender.org/diffusion/B/browse/master/source/blender/nodes/texture/nodes/](https://developer.blender.org/diffusion/B/browse/master/source/blender/nodes/texture/nodes/). Some of these nodes are too high-level (bricks), but the sources could be used to re-create a procedural texture that has been created in Blender and exported to a declarative format. After this proof-of-concept we can define a set of "lowest common denominator" nodes that must be present in an extension.

Another thing we have been talking about is that it would be nice for the generator to be able to use other images/textures in the glTF file as inputs in addition to simple value constants. This way, a noise/mask/grain texture could be embedded into the glTF and then used in the generation step for a procedural wood texture at load time.

I think the focus should be on load-time generation, we don't have to go to runtime evaluation to get the benefits of procedural generation. As long as we take the "pixel-centric" approach, per-pixel evaluation at runtime should be possible in the future.

**Comment by pjoe** [comment]
Right, there are a couple of fundamental building blocks for doing procedural textures, like noise (4D for tiling in uv, 6D if you e.g. also need tiling in time), shape, scatter, etc. My best reference for inspiration is as mentioned Substance Designer, see e.g.: https://www.youtube.com/playlist?list=PLB0wXHrWAmCwWfVVurGIQO_tMVWCFhnqE

For blender source, note that it has a couple of version of nodes: for Cycles and for Eevee. See e.g. 4D perlin noise shader here: https://developer.blender.org/diffusion/B/browse/master/source/blender/gpu/shaders/material/gpu_shader_material_noise.glsl$191

**Comment by pjoe** [comment]
Btw. aslo found my original investigation into tiling noise in Blender: https://blender.stackexchange.com/questions/135437/how-to-make-tileable-procedural-noise-texture

NOTE: 4D noise was added in 2.81 (I think) - not by me :)

**Comment by donmccurdy** [comment]
> After this proof-of-concept we can define a set of "lowest common denominator" nodes that must be present in an extension.

Just to confirm: you're saying this extension is a proof of concept and won't use "lowest common denominator" nodes, but that you think that could be a next step afterward? The latter is obviously a big project, so we can certainly start a new thread for that if you'd rather keep this issue focused on something simpler.

**Comment by rsahlin** [comment]
> Just to confirm: you're saying this extension is a proof of concept and won't use "lowest common denominator" nodes, but that you think that could be a next step afterward? The latter is obviously a big project, so we can certainly start a new thread for that if you'd rather keep this issue focused on something simpler.

I would say that this issue is serves the purpose of discussing how an extension for procedural texture could be done.
From an Ikea perspective we will first focus on a proof of concept - this will most likely not define the semantics needed to parameterize procedural wood textures - this is simply to be able to move forward and verify parts of the solution, eg the quality and size gains that we can expect.
The second phase will then focus on detailing the semantics needed.

As I see it both phases are relevant for procedural textures - It's just that for the moment our focus is on the first phase.
Does this answer your question @donmccurdy ?

**Comment by donmccurdy** [comment]
Ok, I think so. If you feel that the comments here are getting ahead of your immediate goals feel free to let us know and we can split threads as needed. :)

**Comment by rsahlin** [comment]
You do have a valid point
I think we should keep the first phase as internal to Ikea and have this thread for discussing the semantics/nodestructure for procedurally generated textures. 
This way we (Ikea) can 'go about' our proof of concept and we can discuss here what we should aim for when defining the semantics. :-)

**Comment by fire** [comment]
If I wanted to give a MaterialX material to material mapping with a stage0 approach where one pretends that the materialx definition is a png. 

Did anyone specify material used as png?

Stage 1 would be like gltf variants for the procedural properties.

**Comment by rsahlin** [comment]
@fire Hi and thanks for your comment
I am not sure I understand what you mean by `pretends that the materialx definition is a png`?
Do you mean that the input, the recipe if you will, to the procedural generation should be a png (and not the list of operations needed to perform the image genaration)?

Could you please elaborate?

**Comment by fire** [comment]
My friend Lyuma explains, MaterialX in glTF would allow for a
generalization of "KHR_materials_variants" extension - in other words,
MaterialX would allow for character or object customization to textures
baked at load time.

The standard parameters in the shop example of KHR_materials_variants like
colors would be good for a store.

We can use floats to affect the material output. For example, a float of 0
will generate a texture from SDF nodes to show a triangle and a float of 1
will generate a square which is not only a color.

On Thu, 29 Apr 2021 at 23:11, Richard Sahlin ***@***.***>
wrote:
>
> @fire Hi and thanks for your comment
> I am not sure I understand what you mean by pretends that the materialx
definition is a png?
> Do you mean that the input, the recipe if you will, to the procedural
generation should be a png (and not the list of operations needed to
perform the image generation)?

**Comment by fire** [comment]
A (2D) procedural image generator.

It will not aim at affecting geometry or other parts of the material - just the provide texture input for basecolor, normal, roughness and possibly occlusion.

Yes. All the possible PBR GLTF2 material parameters that are supported.

This is exactly what I want.

**Comment by rsahlin** [comment]
> This is exactly what I want.

Great :-)

We are in the process of starting the second phase of the proof of concept - I will be able to share more in the coming months.
Please be patient as we are still in the analysis phase :-)
